ctl/ctCpuAffinity.cpp

```diff
@@ -0,0 +1,220 @@
+
+Licensed under the Apache License, Version 2.0 (the ""License""); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE, MERCHANTABLITY OR NON-INFRINGEMENT.
+
+See the Apache Version 2.0 License for specific language governing permissions and limitations under the License.
+
+*/
+
+/*
+ * ctCpuAffinity.cpp
+ * Implementation of CPU affinity capability detection and shard -> affinity mapping.
+ */
+
+#include "ctCpuAffinity.hpp"
+
+#include <cstdio>
+
+#include <winsock2.h>
+#include <mstcpip.h>
+#include <windows.h>
+#include <vector>
+#include <algorithm>
+
+namespace ctl
+{
+
+    CpuAffinityInfo QueryCpuAffinitySupport() noexcept
+    {
+        CpuAffinityInfo info{};
+
+        // Processor group and counts
+        info.ProcessorGroupCount = static_cast<uint32_t>(GetActiveProcessorGroupCount());
+        // total logical processors across groups
+        uint64_t total = 0;
+        for (WORD g = 0; g < info.ProcessorGroupCount; ++g)
+        {
+            total += GetActiveProcessorCount(g);
+        }
+        info.LogicalProcessorCount = static_cast<uint32_t>(total);
+
+        // Probe SIO_CPU_AFFINITY support using a temporary UDP socket, if available
+        // WSAStartup is expected to have been called by the application earlier.
+#ifdef SIO_CPU_AFFINITY
+        SOCKET s = WSASocketW(AF_INET, SOCK_DGRAM, IPPROTO_UDP, nullptr, 0, WSA_FLAG_OVERLAPPED);
+        if (s != INVALID_SOCKET)
+        {
+            DWORD bytes = 0;
+            // Per MSDN, SIO_CPU_AFFINITY expects a USHORT processor index (0-based)
+            // as its input buffer. Probe support by passing a small USHORT buffer.
+            USHORT probeProcessorIndex = 0;
+            const int rc = WSAIoctl(s, SIO_CPU_AFFINITY, &probeProcessorIndex, sizeof(probeProcessorIndex), nullptr, 0, &bytes, nullptr, nullptr);
+            info.SupportsCpuAffinityIoctl = (rc == 0);
+            closesocket(s);
+        }
+        else
+        {
+            info.SupportsCpuAffinityIoctl = false;
+        }
+#else
+        // SDK doesn't define SIO_CPU_AFFINITY; treat as not supported.
+        info.SupportsCpuAffinityIoctl = false;
+#endif
+
+        // Feature-detect GetQueuedCompletionStatusEx presence conservatively
+        // Assume present on supported Windows versions; leave false only if very old.
+        info.SupportsGQCSEx = true;
+
+        return info;
+    }
+
+    namespace {
+        // Helper: get per-group processor counts
+        static std::vector<uint32_t> GetProcessorCountsPerGroup() noexcept
+        {
+            std::vector<uint32_t> counts;
+            const WORD groupCount = static_cast<WORD>(GetActiveProcessorGroupCount());
```

Comment:
GetActiveProcessorGroupCount returns a WORD, so we don't need a cast right?

counts.reserve(GetActiveProcessorGroupCount());


?